### PR TITLE
Validates the value of the environment variable

### LINKED ISSUE #102732: Document what happens when using envFrom with invalid values in secret/configmap
#### What happened:

Trying to use `envFrom` to set environment variables from a secret with binary data. Got error message:
`Error: failed to start container "mortenlj-test": Error response from daemon: OCI runtime create failed: container_linux.go:349: starting container process caused "process_linux.go:449: container init caused \"setenv: invalid argument\"": unknown`

#### What you expected to happen:

I was hoping that invalid values would be skipped with an event, the same as invalid keys.

#### How to reproduce it (as minimally and precisely as possible):

Create a secret with binary data, specifically NULL-bytes, in the *value*. Try to use this in a pod using `envFrom`.

#### Anything else we need to know?:

The specific problem is mentioned in https://github.com/kubernetes/kubernetes/issues/89906, but a mention in the docs for the `envFrom` field on [Container](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#container-v1-core) would be very useful. The current documentation mentions what happens if the keys are invalid, but makes no mention of what happens if a value is invalid (contains a NULL-byte).

This is mostly a request for clearer documentation, but even better if keys with invalid values was just skipped, the same as if the key itself was invalid.

#### Environment:
- Kubernetes version (use `kubectl version`): 
  Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.17", GitCommit:"68b4e26caf6ede7af577db4af62fb405b4dd47e6",  
  GitTreeState:"clean", BuildDate:"2021-03-18T01:02:41Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
  Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.17-gke.1900", GitCommit:"c97b098770a3dcd2e36fdccd8ec5642a5cddbd7f", 
  GitTreeState:"clean", BuildDate:"2021-04-14T09:22:16Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration: N/A
- OS (e.g: `cat /etc/os-release`): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Network plugin and version (if this is a network-related bug): N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by wzshiming**
#### What type of PR is this?
/kind bug
/sig node

#### What this PR does / why we need it:

Except for Plan9, other systems do not allow null(\x00) in environment variables because null(\x00) is the end of the C String. 
This is mostly for clearer event, and keys with invalid values was skipped, the same as if the key itself was invalid.

#### Which issue(s) this PR fixes:

Fixes #102732

#### Special notes for your reviewer:

Unix `Setenv` https://github.com/golang/go/blob/27f83723e98d8e3795a07bdca2b3a8155b0d72b3/src/syscall/env_unix.go#L97-L130

Windows `Setenv` https://github.com/golang/go/blob/27f83723e98d8e3795a07bdca2b3a8155b0d72b3/src/syscall/env_windows.go#L32-L46

#### Does this PR introduce a user-facing change?

```release-note
Validates the value of the environment variable
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go [hunk 1]
```diff
@@ -415,6 +416,26 @@ func IsEnvVarName(value string) []string {
...
+	// on Plan 9, null is used as a separator, eg in $path.
+	if runtime.GOOS == "plan9" {
+		return errs
+	}
+	for i := 0; i < len(value); i++ {
+		if value[i] == 0 {
+			errs = append(errs, "a valid value must can't contain null")
+			break
+		}
+	}
```
Comment (kolyshkin) [inline staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go:435]: You can simplify this by using e.g. `if strings.IndexByte(value, 0) >= 0`

Comment (wzshiming) [reply staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go]: Thanks updated

**Review by adisky — COMMENTED** [review]
@wzshiming I think the [API documentation](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#container-v1-core) also needs to updated, about adding the validation for env values.

**Review by ehashman — COMMENTED** [review]
/triage accepted
/priority important-longterm

Can we please add some unit tests?

**Comment by wzshiming** [comment]
@adisky @ehashman Thanks, Updated.

**Comment by wzshiming** [comment]
PTAL @smarterclayton @ehashman

**Comment by ehashman** [comment]
/remove-lifecycle rotten

I will review but not currently a priority for me

**Pull request closed without merging**